pkg/controllers/flannelmigration/k8s_resources.go

```diff
@@ -0,0 +1,335 @@
+func (d daemonset) WaitForDaemonsetNotFound(k8sClientset *kubernetes.Clientset, namespace string, interval, timeout time.Duration) error {
+	return wait.PollImmediate(interval, timeout, func() (bool, error) {
+		_, err := k8sClientset.AppsV1().DaemonSets(namespace).Get(string(d), metav1.GetOptions{})
+		if apierrs.IsNotFound(err) {
+			return true, nil
+		}
+		if err != nil {
+			return true, err
+		}
+		// Daemoset still exists, retry.
+		return false, nil
+	})
+}
+
+// Add node selectors to daemonset.
+// If node selectors has been set already, do nothing.
+func (d daemonset) AddNodeSelector(k8sClientset *kubernetes.Clientset, namespace string, selector map[string]string) error {
+	ds, err := k8sClientset.AppsV1().DaemonSets(namespace).Get(string(d), metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	needUpdate := false
+	for k, v := range selector {
+		if currentVal, ok := ds.Spec.Template.Spec.NodeSelector[k]; ok && currentVal == v {
+			continue
+		}
+		ds.Spec.Template.Spec.NodeSelector[k] = v
+		needUpdate = true
+	}
+
+	if needUpdate {
+		_, err = k8sClientset.AppsV1().DaemonSets(namespace).Update(ds)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// k8spod holds a collection of helper functions for Kubernetes pod.
+type k8spod string
+
+// Run a pod with a host path volume on specified node. Wait till it is completed.
+// Return error with log if for any reason pod not completed successfully.
+func (p k8spod) RunPodOnNodeTillComplete(k8sClientset *kubernetes.Clientset, namespace, imageName, nodeName, shellCmd, hostPath string, privileged, hostNetwork bool) (string, error) {
+	podName := string(p)
+	containerName := podName
+	hostPathDirectory := v1.HostPathDirectory
+
+	log.Infof("Create pod on node %s to run [ %s ].", nodeName, shellCmd)
+	podSpec := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: podName + "-",
+			Labels: map[string]string{
+				"flannel-migration": nodeName,
+			},
+		},
+		Spec: v1.PodSpec{
+			Containers: []v1.Container{
+				{
+					Name:    containerName,
+					Image:   imageName,
+					Command: []string{"/bin/sh"},
+					Args: []string{
+						"-c",
+						shellCmd,
+					},
+					VolumeMounts: []v1.VolumeMount{
+						{
+							Name:      "host-dir",
+							MountPath: fmt.Sprintf("/host/%s", hostPath),
+						},
+					},
+					SecurityContext: &v1.SecurityContext{Privileged: &privileged},
+				},
+			},
+			HostNetwork:     hostNetwork,
+			NodeName:        nodeName,
+			RestartPolicy:   v1.RestartPolicyNever,
+			Volumes: []v1.Volume{
+				{
+					Name: "host-dir",
+					VolumeSource: v1.VolumeSource{
+						HostPath: &v1.HostPathVolumeSource{
+							Path: hostPath,
+							Type: &hostPathDirectory,
+						},
+					},
+				},
+			},
+		},
+	}
+
+	logs := ""
+	pod, err := k8sClientset.CoreV1().Pods(namespace).Create(podSpec)
+	if err != nil {
+		return logs, err
+	}
+
+	err = waitForPodSuccessTimeout(k8sClientset, pod.Name, pod.Namespace, 1*time.Second, 2*time.Minute)
```

Comment:
I feel like 2 minutes might not be quite long enough here for some scenarios where, e.g., it might take a substantial amount of time to pull the image or to get the pod running in the first place.

Reply:
Agree.Extended to 5 minutes.